library/ssl_tls.c

```diff
@@ -9177,8 +9177,9 @@ int mbedtls_ssl_conf_psk( mbedtls_ssl_config *conf,
     ssl_conf_remove_psk( conf );
 
     /* Check and set raw PSK */
-    if( psk == NULL || psk_len > MBEDTLS_PSK_MAX_LEN )
```

Comment:
I am not sure we should disallow a PSK that is both `NULL` and `0` length. We'll need another reviewer to comment.

Reply:
I was also thinking about it. On the other head if we allow psk_len with 0 then line 9182 will return error. To protect it we can shift checking psk_len before calloc

Reply:
We need to support zero-length PSKs if we want to be strictly compliant:
> TLS implementations supporting these ciphersuites MUST support arbitrary PSK identities up to 128 octets in length, and arbitrary PSKs up to 64 octets in length.

On the other hand I can't imagine any valid use case for it.

Reply:
We were blocking `NULL` PSK before, so I don't see how we would have supported zero-length PSK previously. One would expect `NULL` for PSK and `0` for its length would be the way to specify a zero-length PSK. This would be rejected before this PR and would be rejected afterwards. This PR doesn't appear to reduce our compliance further.

Reply:
Many APIs that take a buffer and size as arguments expect a non-null pointer even when the size is 0, for example standard C functions like `memset` and `memcpy`. I don't know if we supported a zero-length PSK before, but if we didn't, it wasn't solely because of this null check.

In TLS, when a ciphersuite uses a pre-shared key, that key is always on the critical path for authentication. So I don't see any use case for a PSK that is short enough to be guessed, let alone 0. (It could be different for other protocols, e.g. [NoisePSK](https://moderncrypto.org/mail-archive/noise/2015/000393.html) where the PSK provides additional security, unlike TLS where it's always necessary for security.) So I'm comfortable with not supporting a zero-length PSK.